### PR TITLE
Add async write

### PR DESCRIPTION
**Pull request by MingweiSamuel**
depends on #168

### CODE REVIEW

File: hydroflow_lang/src/graph/mod.rs [hunk 1]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
-                // A <- H.
-                new_preds[hoff_id].insert(zero_index, (src, *src_idx));
-                // H <- Z.
-                new_preds[dst].insert(*dst_idx, (hoff_id, zero_index));
-                // H -> Z.
-                new_succs[hoff_id].insert(zero_index, (dst, *dst_idx));
+    // Determine node's subgraph and subgraph's nodes.
+    // This list of nodes in each subgraph are to be in topological sort order.
+    let (node_subgraph, subgraph_nodes) = {
+        // Ignore handoffs. Esentially each subgraph has it's own topo sort.
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:248]: Ignore handoffs for now. First ensure that each subgraph has its own topo sort.

File: hydroflow_lang/src/graph/mod.rs [hunk 2]
```diff
@@ -1,19 +1,25 @@
 //! Graph representation stages for Hydroflow graphs.
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:1]: Not reviewing this as closely on the assumption that you've postponed handling some of the comments from previous PR.

File: hydroflow_lang/src/graph/ops/mod.rs [hunk 3]
```diff
@@ -9,9 +9,18 @@ use syn::{Expr, GenericArgument, Token};
 
 use super::{GraphNodeId, GraphSubgraphId};
 
+#[derive(Clone, Copy, PartialOrd, Ord, PartialEq, Eq, Debug)]
+pub enum InputBarrier {
+    None,
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/mod.rs:14]: remove

File: hydroflow_lang/src/graph/ops/mod.rs [hunk 4]
```diff
@@ -9,9 +9,18 @@ use syn::{Expr, GenericArgument, Token};
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/ops/mod.rs:9]: not reviewing this carefully on the assumption that changes from the previous PR review are being postponed

File: hydroflow_lang/src/graph/partitioned_graph.rs [hunk 5]
```diff
@@ -168,30 +170,44 @@ impl PartitionedGraph {
...
                         let pull_ident =
                             self.node_id_as_ident(subgraph_nodes[pull_to_push_idx - 1], false);
+
                         let push_ident =
-                            self.node_id_as_ident(subgraph_nodes[pull_to_push_idx], false);
+                            if let Some(&node_id) = subgraph_nodes.get(pull_to_push_idx) {
+                                self.node_id_as_ident(node_id, false)
+                            } else {
+                                // Entire subgraph is pull.
+                                assert_eq!(
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/partitioned_graph.rs:187]: Is there a test that triggers this case?

**Review by jhellerstein — APPROVED** [review]

File: hydroflow/tests/surface_codegen.rs [hunk 6]
```diff
@@ -219,3 +368,72 @@ pub fn test_covid_tracing() {
...
+    df.run_available();
+
+    println!("B");
+
+    pairs_send.send((0, 3)).unwrap();
+    pairs_send.send((0, 3)).unwrap();
+    df.run_available();
+}
+
+#[tokio::test(flavor = "current_thread")]
```
[the remark this answers is not in the document]
Comment (MingweiSamuel) [reply hydroflow/tests/surface_codegen.rs:412]: Just either "current_thread" or "multi_thread"

File: hydroflow/tests/surface_stratum.rs [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+//     let mut df = hydroflow_syntax! {
+//         a = identity::<usize>();
+//         a -> a;
+//     };
+// }
+
+#[test]
+pub fn test_difference_a() {
+    let output = <Rc<RefCell<Vec<usize>>>>::default();
+    let output_inner = output.clone();
```
[the remark this answers is not in the document]
Comment (MingweiSamuel) [reply hydroflow/tests/surface_stratum.rs:20]: Increments the RC, changed to `Rc::clone(&output)` to be clearer

File: hydroflow_lang/src/graph/partitioned_graph.rs (hunk 5 above)
Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/partitioned_graph.rs:187]: Yeah a few. Seems to hit this case pretty often when you're pulling from one or more sources and pushing to a single handoff (the handoff "push" doesn't count when I say "Entire subgraph is pull.")

**Pull request merged**